mods/default/functions.lua

```diff
@@ -302,79 +295,78 @@ default.after_place_leaves = function(pos, placer, itemstack, pointed_thing)
+-- Spread the load of finding trunks
+default.leafdecay_trunk_find_allow_accumulator = 0
+
+-- Finds_per_second = 5000
+minetest.register_globalstep(function(dtime)
+	default.leafdecay_trunk_find_allow_accumulator = math.floor(dtime * 5000)
+end)
+
 minetest.register_abm({
+	label = "Leaf decay",
 	nodenames = {"group:leafdecay"},
-	neighbors = {"air", "group:liquid"},
-	-- A low interval and a high inverse chance spreads the load
+	neighbors = {"air"},
 	interval = 2,
 	chance = 5,
+	catch_up = false,
 
 	action = function(p0, node, _, _)
-		--print("leafdecay ABM at "..p0.x..", "..p0.y..", "..p0.z..")")
-		local do_preserve = false
-		local d = minetest.registered_nodes[node.name].groups.leafdecay
-		if not d or d == 0 then
-			--print("not groups.leafdecay")
-			return
-		end
-		local n0 = minetest.get_node(p0)
-		if n0.param2 ~= 0 then
-			--print("param2 ~= 0")
+		if node.param2 ~= 0 then
 			return
 		end
-		local p0_hash = nil
-		if default.leafdecay_enable_cache then
-			p0_hash = minetest.hash_node_position(p0)
-			local trunkp = default.leafdecay_trunk_cache[p0_hash]
-			if trunkp then
-				local n = minetest.get_node(trunkp)
-				local reg = minetest.registered_nodes[n.name]
-				-- Assume ignore is a trunk, to make the thing
-				-- work at the border of the active area
-				if n.name == "ignore" or (reg and reg.groups.tree and
-						reg.groups.tree ~= 0) then
-					--print("cached trunk still exists")
-					return
-				end
-				--print("cached trunk is invalid")
-				-- Cache is invalid
-				table.remove(default.leafdecay_trunk_cache, p0_hash)
+
+		local p0_hash = minetest.hash_node_position(p0)
+		local trunkp = default.leafdecay_trunk_cache[p0_hash]
+		if trunkp then
+			local trunkn = minetest.get_node(trunkp)
+			local reg = minetest.registered_nodes[trunkn.name]
+			-- Assume ignore is a trunk, to make the thing
+			-- work at the border of the active area
+			if trunkn.name == "ignore" or (reg and reg.groups.tree) then
+				-- Cached trunk still exists
+				return
 			end
+			-- Cache is invalid
+			default.leafdecay_trunk_cache[p0_hash] = nil
 		end
+
 		if default.leafdecay_trunk_find_allow_accumulator <= 0 then
 			return
 		end
 		default.leafdecay_trunk_find_allow_accumulator =
-				default.leafdecay_trunk_find_allow_accumulator - 1
+			default.leafdecay_trunk_find_allow_accumulator - 1
+
+		local d = minetest.registered_nodes[node.name].groups.leafdecay
 		-- Assume ignore is a trunk, to make the thing
 		-- work at the border of the active area
-		local p1 = minetest.find_node_near(p0, d, {"ignore", "group:tree"})
-		if p1 then
-			do_preserve = true
-			if default.leafdecay_enable_cache then
-				--print("caching trunk")
-				-- Cache the trunk
-				default.leafdecay_trunk_cache[p0_hash] = p1
-			end
+		if #minetest.find_nodes_in_area(
+				vector.subtract(p0, d),
+				vector.add(p0, d),
+				{"ignore", "group:tree"}) > 0 then
+			-- Cache the trunk
+			default.leafdecay_trunk_cache[p0_hash] = true
```

Comment:
The value it is used in line 325 by `get_node()`, with `true` it doesn't actually cache anything right now.